trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -0,0 +1,213 @@
+   *   array defining the specific header and the headers must be listed in order!
+   *   Each header item must consist of a header name (key: 'name'), and type (key: 'type',
+   *   supported values: 'required', 'optional').
+   * 
+   *   This list of headers is expected to appear in the data file header row.
+   *   
+   *   NOTE: the set headers array is zero-based index and represents the order
+   *         of the header as each item appears in the headers array parameter
+   *         and is unaltered by setters and getters.
+   * 
+   * @throws \Exception
+   *  - An empty array header.
+   * 
+   * @return void
+   */
+  public function setHeaders(array $headers) {
+
+    // Headers must have a value.
+    if (empty($headers)) {
+      throw new \Exception('The Headers Trait requires an array of headers and must not be empty.');  
+    }
+    
+    // Required keys that each header element must posses and 
+    // could not be set to an empty value. Type key value must be
+    // one of the type values defined by the types property. 
+    $required_header_keys = [
+      'name', // Name of the header.
+      'type'  // Type of the header (ie. required or optional).
+    ];
+    
+    // Each header, check required keys and value and place 
+    // in designated type array of the context.
+    $context_headers = [];
+
+    foreach($headers as $index => $header) {
+      // Header element key and value check.
+      foreach($required_header_keys as $key) {
+        // Key is not set.
+        if (!isset($header[ $key ])) {
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' when defining headers.');
+        }
+        
+        // Key is set but value is empty.
+        if (empty(trim($header[ $key ]))) {
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' to be have a value.');
+        }
+
+        // Type value is one of valid types.
+        if ($key == 'type' && !in_array($header[ $key ], $this->types)) {
+          $str_types = implode(', ', $this->types);
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' value to be one of [' . $str_types . '].');
+        }
+      }
+
+      // With the header type already verified to be one of the valid types, 
+      // push the header into the designated type temporary context array.
+      $context_headers[ $header['type'] ][ $index ] = $header['name'];
+    }
+    
+    // Set each header type context array.
+    foreach($this->types as $type) {  
+      $this->context[ $this->context_key ][ $type ] = $context_headers[ $type ] ?? [];
+    }
+  }
+  
+  /**
+   * Get required headers.
+   * 
+   * @return array
+   *   All headers of type required, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key required in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
```

Comment:
I removed `The key required in the context headers array set by the setter method.` because the "key" mentioned here I assume refers to the context or type key, but this is not obvious when the previous paragraph talks about the array being keyed by index. I don't think we need to mention the context key at all since we are supposed to be talking about the return value here, so I suggest just truncating the two paragraphs into one:

```suggestion
   *   All headers of type 'required', keyed by the index (order) from
   *   the headers array and header name as the value.
   *   NOTE: the array is zero-based indexed.
```